source-map.bs

```diff
@@ -497,13 +497,10 @@ sources=] <dfn for="decode source map mappings">|sources|</dfn>, run the followi
             |groups|[|generatedLine|] on `,`.
         1. Let |generatedColumn| be 0.
         1. [=For each=] |segment| in |segments|:
-            1. Let |position| be a [=position variable=] for |segment|, initially pointing at
-                |segment|'s start.
-            1. [=Decode a base64 VLQ=] from |segment| given |position| and let
-                |relativeGeneratedColumn| be the result.
-            1. If |relativeGeneratedColumn| is null, [=optionally report an error=] and continue
-                with the next iteration.
-            1. Increase |generatedColumn| by |relativeGeneratedColumn|. If the result is negative,
+            1. [=Decode a list of base64 VLQ=] from |segment| and let |vlqs| be the result.
+            1. If |vlqs|'s [=list/size=] is not 1, 4 or 5, [=optionally report an error=]
+                and continue with the next iteration.
```

Comment:
If the length is, for example, 2, do you know if iterations completely skip the segnemt or do they still add the value of the first VLQ to _generatedColumn_?

Reply:
Great point, Chrome DevTools [throws](https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/core/sdk/SourceMap.ts;l=550;drc=81f47ed35b4c6d4e7ddc5aa538d056b1c37a364e) in the case where `vlqs` has a size of 2. It would add a mapping first, but then aborts decoding mappings via a throw, because it expects being able to read 3 more VLQs at least.

Firefox does [the same](https://github.com/mozilla/source-map/blob/master/wasm-mappings/src/lib.rs#L660) in their wasm implementation.

WebKit inspector is somewhat undefined as they'll start [doing math on `undefined`](https://github.com/WebKit/WebKit/blob/e4c64c9c92bd4bad8d1dfe9615d6b5dae9de00dc/Source/WebInspectorUI/UserInterface/Models/SourceMap.js#L249) in their VLQ decoding code, which I assume will result in some `NaN`s.

VSCode uses `@jrdigewell/sourcemap-codec` under the hood which handles it [similar to WebKit](https://github.com/jridgewell/sourcemap-codec/blob/main/src/sourcemap-codec.ts#L45).

All in all there seems to be a split between either throwing or converting to NaNs. (To be fair I only skimmed all of these so I might have missed validation steps).

From a spec perspective I think it's ok that we reject segments with an invalid amount of VLQ numbers but I don't have a strong opinion on this.

Reply:
> VSCode uses `@jrdigewell/sourcemap-codec` under the hood which handles it [similar to WebKit](https://github.com/jridgewell/sourcemap-codec/blob/main/src/sourcemap-codec.ts#L45).

Codec'll produce a segment of length 4 filling with 0s. I don't perform any validation, checking wastes time and we can just assume maps are well formed because they're not user input.

And, VSCode uses sourcemap-codec? 😄

Reply:
An alternative is to return a `[=queue=]` instead of a `[=list=]` and consume values via `dequeue`. This would also fit better with the iterator style that most implementations employ.

Reply:
What about:
- capturing the length of the list as soon as we get it
- fill the list with zeroes

```
- Let |list| be ...
- Let |count| be |list|'s length
- Append as many 0 as needed at the end of |list|, so that |list|'s |length| is at least 5.
```

Reply:
And then compare against `|count|` going forward to correctly handle null mappings or mappings without a name? I think then I'd prefer

```
1. Increase |sourceIndex| by |vlqs[1]|.
1. If |vlqs[2]| [=list/exists=], increase |originalLine| by |vlqs[2]|.
1. If |vlqs[3]| [=list/exists=], increase |originalColumn| by |vlqs[3]|.
```

That would be equivalent to filling with zeroes.

Reply:
This seems good too.

Reply:
Ok, I removed the check for length 1, 4 or 5, and replaced it with the conditional increase. I also fixed the formatting for variables (`|..|`). PT(another)L